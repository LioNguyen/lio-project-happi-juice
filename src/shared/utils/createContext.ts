import type React from "react";
import { createContext as originCreateContext, useContext } from "react";

function createContext<T>(): [React.Provider<T>, () => T, React.Context<T>] {
  const context = originCreateContext<T>({} as T);

  return [context.Provider, () => useContext(context), context];
}

export { createContext };
