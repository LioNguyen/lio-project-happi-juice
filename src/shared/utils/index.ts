export * from './lazyLoad'
export * from './tailwind'
