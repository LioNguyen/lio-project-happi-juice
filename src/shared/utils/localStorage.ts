/**
 * Save data to localStorage
 * @param key - The key under which the data will be stored
 * @param value - The data to store, can be any serializable value
 */
function setLocalStorage<T>(key: string, value: T): void {
  try {
    const serializedValue =
      typeof value === 'string' ? value : JSON.stringify(value)

    localStorage.setItem(key, serializedValue)
  } catch (error) {
    console.error(`Error saving ${key} to localStorage:`, error)
  }
}

/**
 * Retrieve data from localStorage
 * @param key - The key of the stored data
 * @returns The parsed data or null if not found or error occurs
 */
function getLocalStorage<T>(key: string): T | null {
  try {
    const serializedValue = localStorage.getItem(key)

    if (!serializedValue || serializedValue === 'undefined') {
      return null
    }

    // Try to parse JSON
    try {
      return JSON.parse(serializedValue) as T
    } catch {
      // Return as string if JSON parsing fails
      return serializedValue as T
    }
  } catch (error) {
    console.error(`Error reading ${key} from localStorage:`, error)
    return null
  }
}

/**
 * Remove data from localStorage
 * @param key - The key to remove from localStorage
 */
function removeLocalStorage(key: string): void {
  try {
    localStorage.removeItem(key)
  } catch (error) {
    console.error(`Error removing ${key} from localStorage:`, error)
  }
}

/**
 * Clear all data from localStorage
 */
function clearLocalStorage(): void {
  try {
    localStorage.clear()
  } catch (error) {
    console.error('Error clearing localStorage:', error)
  }
}

export {
  clearLocalStorage,
  getLocalStorage,
  removeLocalStorage,
  setLocalStorage,
}
