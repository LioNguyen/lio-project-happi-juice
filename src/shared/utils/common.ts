import { format, isValid, Locale, parseISO } from 'date-fns'
import { enUS, vi } from 'date-fns/locale'
import _ from 'lodash'

import { IOrderItem } from '@/domains/order'
import { ISelectOption } from '@/shared/types'

interface INumberFormatOptions {
  decimals?: number
  thousandSeparator?: string
  decimalSeparator?: string
}

// ==================== Date Utilities ====================
/**
 * Parse various date formats into Date object
 * @category Date
 * @param {Date | string | undefined} value - Input date
 * @returns {Date | undefined} Parsed Date object or undefined if invalid
 * @example
 * parseDate("2024-02-26") // Date object
 * parseDate("invalid") // undefined
 */
function parseDate(value: Date | string | undefined): Date | undefined {
  if (!value) return undefined
  if (value instanceof Date) return isValid(value) ? value : undefined

  try {
    // Try ISO string parsing first
    const date = parseISO(value)
    if (isValid(date)) return date

    // Fallback to direct Date creation
    const normalDate = new Date(value)
    return isValid(normalDate) ? normalDate : undefined
  } catch {
    return undefined
  }
}

/**
 * Format a date string according to specified format
 * @category Date
 * @param {string} date - Date string to format
 * @param {string} [dateFormat] - Output format (defaults to "yyyy-MM-dd")
 * @returns {string} Formatted date string
 * @example
 * formatDateTime("2024-02-26") // "2024-02-26"
 * formatDateTime("2024-02-26", "dd/MM/yyyy") // "26/02/2024"
 */
function formatDateTime(date: string, dateFormat?: string) {
  return format(date, dateFormat || 'yyyy-MM-dd')
}

const localeMap: Record<string, Locale> = {
  vi: vi,
  en: enUS,
}

// ==================== Number & Price Utilities ====================
/**
 * Format a number with customizable separators and decimals
 * @category Number
 * @param {number} value - Number to format
 * @param {INumberFormatOptions} [options] - Formatting options
 * @returns {string} Formatted number string
 */
const formatNumber = (
  value: number,
  options?: INumberFormatOptions,
): string => {
  const {
    decimals = 0,
    thousandSeparator = ',',
    decimalSeparator = '.',
  } = options || {}

  const parts = value.toFixed(decimals).split('.')
  parts[0] = parts[0].replace(/\B(?=(\d{3})+(?!\d))/g, thousandSeparator)

  return parts.join(decimalSeparator)
}

/**
 * Formats the price with thousand separators using locale-specific formatting
 * @category Price
 * @param {string} price - Price string to format
 * @returns {string} Formatted price string
 */
const formatPrice = (price: string) => {
  return Number(price).toLocaleString()
}

// ==================== Array Utilities ====================
/**
 * Transforms an array of select options into a sorted, joined string
 * @category Array
 * @param {ISelectOption[]} array - Array of select options
 * @param {string} [key] - Key to extract from each option
 * @param {string} [separator] - Separator for joining values
 * @returns {string} Sorted and joined string
 */
const transformToSortedString = (
  array: ISelectOption[],
  key: string = 'value',
  separator: string = ',',
): string => {
  return _.chain(array).map(key).sortBy().join(separator).value()
}

// ==================== Order Utilities ====================
/**
 * Calculates the total price for an array of order items
 * @category Order
 * @param {IOrderItem[]} items - Array of order items
 * @returns {number} Total price
 */
const calculateTotalPrice = (items: IOrderItem[]) => {
  return items.reduce((sum, item) => sum + item.price * item.quantity, 0)
}

export {
  // Date utils
  parseDate,
  formatDateTime,
  localeMap,

  // Number & Price utils
  formatNumber,
  formatPrice,

  // Array utils
  transformToSortedString,
  // Order utils
  calculateTotalPrice,
}
