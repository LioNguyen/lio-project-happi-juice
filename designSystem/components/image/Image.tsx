import { type FC, type HTMLAttributes, memo, useState } from "react";

import { IComponentBase } from "@/shared/types";
import { cn } from "@/shared/utils";

interface IImageProps extends IComponentBase {
  src: string;
  alt?: string;
  size?: number | string;
  imgProps?: Omit<HTMLAttributes<HTMLImageElement>, "src" | "alt" | "onClick">;
  onClick?: () => void;
}

const Image: FC<IImageProps> = ({ className, src, alt, size, imgProps, onClick }) => {
  const [isImageLoaded, setIsLoaded] = useState(false);
  const { className: imgClassName, ...restImgProps } = imgProps || {};

  const handleLoad = () => {
    setIsLoaded(true);
  };

  return (
    <div data-testid="image-test" className={cn("image", onClick ? "cursor-pointer" : "", className)} onClick={onClick}>
      <img
        className={cn("transition-opacity duration-300", imgClassName, isImageLoaded ? "block" : "hidden")}
        src={src}
        alt={alt || src}
        width={size}
        height={size}
        onLoad={handleLoad}
        {...restImgProps}
      />
    </div>
  );
};

export default memo(Image);
export type { IImageProps };
