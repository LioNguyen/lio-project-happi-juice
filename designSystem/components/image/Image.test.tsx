import { fireEvent, screen } from "@testing-library/react";
import { ComponentPropsWithRef } from "react";
import { vi } from "vitest";

import { renderWithTheme } from "@/shared/utils";
import Image from "./Image";

type ImagePropsWithRef = ComponentPropsWithRef<typeof Image>;

const defaultProps = {
  src: "./logo-path.svg",
  alt: "logo",
};

const renderComponent = (props: Partial<ImagePropsWithRef>) => {
  return renderWithTheme(<Image {...defaultProps} {...props} />);
};

describe("Image test suite", () => {
  // Snapshot Tests
  it("should match snapshot with default props", () => {
    const { container } = renderComponent({});
    expect(container).toMatchSnapshot();
  });

  it("should render correctly", () => {
    renderComponent({});

    const img = screen.getByAltText(defaultProps.alt);
    expect(img).toBeInTheDocument();
    expect(img.closest("div")).toBeInTheDocument();
    expect(img.closest("div")).toBeInTheDocument();
  });

  it("should render class correctly", () => {
    const className = "custom-class";
    renderComponent({ className });

    const img = screen.getByAltText("logo");
    expect(img.closest("div")).toHaveClass("image");
    expect(img.closest("div")).toHaveClass(className);
  });

  it("should render src correctly", async () => {
    const src = "custom-src";
    renderComponent({ src });

    const img = screen.getByAltText(defaultProps.alt);

    // Trigger load event
    fireEvent.load(img);

    expect(img).toBeInTheDocument();
    expect(img).toHaveAttribute("src", src);
    expect(img).toHaveClass("block");
    expect(img).not.toHaveClass("hidden");
  });

  it("should render no image when src is error -> onLoad not run", () => {
    const src = "";
    renderComponent({ src });

    const img = screen.getByAltText(defaultProps.alt);

    expect(img).toBeInTheDocument();
    expect(img).toHaveAttribute("src", src);
    expect(img).toHaveClass("hidden");
    expect(img).not.toHaveClass("block");
  });

  it("should render alt correctly when alt prop is falsy", () => {
    const alt = "";
    const src = "src-test";
    renderComponent({ alt, src });

    const img = screen.getByAltText(src);
    expect(img).toHaveAttribute("alt", src);
  });

  it("should render size correctly", () => {
    const size = "25px";
    renderComponent({ size });

    const img = screen.getByAltText(defaultProps.alt);
    expect(img).toBeInTheDocument();
    expect(img).toHaveAttribute("width", size);
    expect(img).toHaveAttribute("height", size);
  });

  it("should handle onClick correctly", () => {
    const onClick = vi.fn();
    renderComponent({ onClick });

    const image = screen.getByTestId("image-test");
    fireEvent.click(image);

    expect(image).toHaveClass("cursor-pointer");
    expect(onClick).toHaveBeenCalledTimes(1);
  });

  it("should render imgProps correctly", () => {
    const imgProps = {
      className: "w-full",
    };
    renderComponent({ imgProps });

    const img = screen.getByAltText(defaultProps.alt);
    expect(img).toHaveClass(imgProps.className);
  });
});
