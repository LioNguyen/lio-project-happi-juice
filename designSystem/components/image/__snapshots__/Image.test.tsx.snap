// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Image test suite > should match snapshot with default props 1`] = `
<div>
  <div
    class="image"
    data-testid="image-test"
  >
    <img
      alt="logo"
      class="transition-opacity duration-300 hidden"
      src="./logo-path.svg"
    />
  </div>
</div>
`;
